Track pods by UID instead of only namespace/name

saleemrashid opened the pull request.

twelho committed:
fix: do not register `nil` loggers

`nopSpan.Logger()` passes in `nil` into `log.WithLogger` any time `trace.StartSpan` is invoked, which overrides the logger from an externally provided context. This inadvertently suppresses all logging for the span, including for errors. Prevent clearing the existing logger here if, for example, `trace.StartSpan` does not pass in a new one. To use the default logger again, invoke `log.WithLogger` with `log.L` explicitly.

Signed-off-by: Dennis Marttinen <twelho@welho.tech>

thevilledev committed:
fix: enable lll linter and format long function signatures

Previously lll linter was configured to .golangci.yml but with
a typo in the directive ('linter-settings', not 'linters-settings').

Enable line length linter and fix long lines by breaking function
signatures across multiple lines in exec.go, attach.go, and resource.go.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

saleemrashid committed:
queue: Use generic key type

saleemrashid committed:
node: Add PodUIDLifecycleHandler

saleemrashid committed:
node: Track pods by UID instead of only namespace/name

saleemrashid committed:
node: Implement PodUIDLifecycleHandler for mockProvider

senyosimpson commented on node/podcontroller.go:
```diff
@@ -67,12 +77,22 @@ type PodLifecycleHandler interface {
 	// concurrently outside of the calling goroutine. Therefore it is recommended
 	// to return a version after DeepCopy.
-	GetPods(context.Context) ([]*corev1.Pod, error)
+	GetPodByUID(ctx context.Context, namespace, name string, uid types.UID) (*corev1.Pod, error)
```
why don't we just add these to the PodLifecycleHandler directly? or do you want some type safety against calling GetPods/GetPodStatus?

saleemrashid replied:
For the type safety. I want this to be backwards-compatible (I'm going to see if we can get it upstreamed), but I also don't want the old methods to be accidentally called internally.

senyosimpson reviewed (approved):
lgtm

galal-hussein committed:
Check for EphemeralContainers changes in pod compare

Signed-off-by: galal-hussein <hussein.galal.ahmed.11@gmail.com>

pires committed:
ci: revisit Go linter

dependabot[bot] committed:
build(deps): bump golang.org/x/net from 0.34.0 to 0.38.0

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.34.0 to 0.38.0.
- [Commits](https://github.com/golang/net/compare/v0.34.0...v0.38.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-version: 0.38.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump github.com/prometheus/common from 0.61.0 to 0.63.0

Bumps [github.com/prometheus/common](https://github.com/prometheus/common) from 0.61.0 to 0.63.0.
- [Release notes](https://github.com/prometheus/common/releases)
- [Changelog](https://github.com/prometheus/common/blob/main/RELEASE.md)
- [Commits](https://github.com/prometheus/common/compare/v0.61.0...v0.63.0)

---
updated-dependencies:
- dependency-name: github.com/prometheus/common
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

enrichman committed:
added envtest

dependabot[bot] committed:
build(deps): bump go.opentelemetry.io/otel from 1.33.0 to 1.36.0

Bumps [go.opentelemetry.io/otel](https://github.com/open-telemetry/opentelemetry-go) from 1.33.0 to 1.36.0.
- [Release notes](https://github.com/open-telemetry/opentelemetry-go/releases)
- [Changelog](https://github.com/open-telemetry/opentelemetry-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/open-telemetry/opentelemetry-go/compare/v1.33.0...v1.36.0)

---
updated-dependencies:
- dependency-name: go.opentelemetry.io/otel
  dependency-version: 1.36.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump go.opentelemetry.io/otel/sdk from 1.33.0 to 1.36.0

Bumps [go.opentelemetry.io/otel/sdk](https://github.com/open-telemetry/opentelemetry-go) from 1.33.0 to 1.36.0.
- [Release notes](https://github.com/open-telemetry/opentelemetry-go/releases)
- [Changelog](https://github.com/open-telemetry/opentelemetry-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/open-telemetry/opentelemetry-go/compare/v1.33.0...v1.36.0)

---
updated-dependencies:
- dependency-name: go.opentelemetry.io/otel/sdk
  dependency-version: 1.36.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

tskillian committed:
Fix potential NPE on pod status updates

dependabot[bot] committed:
build(deps): bump sigs.k8s.io/controller-runtime from 0.19.4 to 0.22.1

Bumps [sigs.k8s.io/controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) from 0.19.4 to 0.22.1.
- [Release notes](https://github.com/kubernetes-sigs/controller-runtime/releases)
- [Changelog](https://github.com/kubernetes-sigs/controller-runtime/blob/main/RELEASE.md)
- [Commits](https://github.com/kubernetes-sigs/controller-runtime/compare/v0.19.4...v0.22.1)

---
updated-dependencies:
- dependency-name: sigs.k8s.io/controller-runtime
  dependency-version: 0.22.1
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump actions/checkout from 4 to 5

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 5.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump golang.org/x/sync from 0.14.0 to 0.17.0

Bumps [golang.org/x/sync](https://github.com/golang/sync) from 0.14.0 to 0.17.0.
- [Commits](https://github.com/golang/sync/compare/v0.14.0...v0.17.0)

---
updated-dependencies:
- dependency-name: golang.org/x/sync
  dependency-version: 0.17.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump go.opentelemetry.io/otel/sdk from 1.36.0 to 1.38.0

Bumps [go.opentelemetry.io/otel/sdk](https://github.com/open-telemetry/opentelemetry-go) from 1.36.0 to 1.38.0.
- [Release notes](https://github.com/open-telemetry/opentelemetry-go/releases)
- [Changelog](https://github.com/open-telemetry/opentelemetry-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/open-telemetry/opentelemetry-go/compare/v1.36.0...v1.38.0)

---
updated-dependencies:
- dependency-name: go.opentelemetry.io/otel/sdk
  dependency-version: 1.38.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump golang.org/x/time from 0.9.0 to 0.13.0

Bumps [golang.org/x/time](https://github.com/golang/time) from 0.9.0 to 0.13.0.
- [Commits](https://github.com/golang/time/compare/v0.9.0...v0.13.0)

---
updated-dependencies:
- dependency-name: golang.org/x/time
  dependency-version: 0.13.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump k8s.io/kubelet from 0.31.4 to 0.34.0

Bumps [k8s.io/kubelet](https://github.com/kubernetes/kubelet) from 0.31.4 to 0.34.0.
- [Commits](https://github.com/kubernetes/kubelet/compare/v0.31.4...v0.34.0)

---
updated-dependencies:
- dependency-name: k8s.io/kubelet
  dependency-version: 0.34.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

pires committed:
ci: bump Kubernetes to v1.33 and Go to 1.24

dependabot[bot] committed:
build(deps): bump github.com/spf13/pflag from 1.0.6 to 1.0.10

Bumps [github.com/spf13/pflag](https://github.com/spf13/pflag) from 1.0.6 to 1.0.10.
- [Release notes](https://github.com/spf13/pflag/releases)
- [Commits](https://github.com/spf13/pflag/compare/v1.0.6...v1.0.10)

---
updated-dependencies:
- dependency-name: github.com/spf13/pflag
  dependency-version: 1.0.10
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

pires committed:
deps: fix Prometheus bump proposed by dependabot

saleemrashid committed:
node: Add comments explaining uidProviderWrapper/syncProviderWrapper backwards compatibility

saleemrashid committed:
node: Add UIDBasedProvider wrapper for legacy PodLifecycleHandler methods
